src/abstractdataframe/selectionfast.jl

```diff
@@ -0,0 +1,152 @@
+                sumz += zi
+            end
+        catch e
+            if e isa MethodError && e.f === zero
+                sumz_undefined = true
+                break
+            else
+                throw(e)
+            end
+        end
+    end
+    # this will happen if eltype of some columns do not support zero
+    # or all columns have eltype Missing
+    sumz_undefined && return nothing
+    init = fill!(Tables.allocatecolumn(typeof(sumz), length(cols[1])), sumz)
+
+    return foldl(cols, init=init) do l, r
+        l .= ifelse.(ismissing.(r), l, l .+ r)
+    end
+end
+
+@noinline function table_transformation(df_sel::AbstractDataFrame, ::ByRow{typeof(length)})
+    @assert ncol(df) > 0
+    fill(ncol(df_sel), nrow(df_sel))
+end
+
+@noinline table_transformation(df_sel::AbstractDataFrame,
+                               ::ByRow{typeof(length∘skipmissing)}) =
+    _length_skipmissing_fast(map(identity, eachcol(df_sel)))
+
+function _length_skipmissing_fast(cols::Vector{<:AbstractVector})
+    @assert !isempty(cols)
+    len = fill(length(cols), length(cols[1]))
+    for col in cols
+        (Missing <: eltype(col)) && (len .-= ismissing.(col))
+    end
+    return len
+end
+
+@noinline function table_transformation(df_sel::AbstractDataFrame, ::typeof(mean))
+    @assert ncol(df_sel) > 0
+    return mean(map(identity, eachcol(df_sel)))
+end
+
+@noinline table_transformation(df_sel::AbstractDataFrame, ::ByRow{typeof(mean)}) =
+    table_transformation(df_sel, mean)
+
+@noinline function table_transformation(df_sel::AbstractDataFrame,
+                                        fun::typeof(ByRow(mean∘skipmissing)))
+    fastmean = _mean_skipmissing_fast(map(identity, eachcol(df_sel)))
+    isnothing(fastmean) || return fastmean
+    slowmean = default_table_transformation(df_sel, fun)
+    isconcretetype(nonmissingtype(eltype(slowmean))) && return slowmean
+    T = mapreduce(typeof, promote_type, slowmean)
+    return convert(AbstractVector{T}, slowmean)
+end
+
+function _mean_skipmissing_fast(cols::Vector{<:AbstractVector})
+    @assert !isempty(cols)
+    local sumz
+    sumz_undefined = true
+    for col in cols
+        try
+            zi = zero(eltype(col))
+            if sumz_undefined
+                sumz_undefined = false
+                sumz = zi
+            elseif !ismissing(zi) # zi is missing if eltype is Missing
+                sumz += zi
+            end
+        catch e
+            if e isa MethodError && e.f === zero
+                sumz_undefined = true
+                break
+            else
+                throw(e)
+            end
+        end
+    end
+    # this will happen if eltype of some columns do not support zero
+    # or all columns have eltype Missing
+    sumz_undefined && return nothing
+    sumv = fill!(Tables.allocatecolumn(typeof(sumz / 0), length(cols[1])), sumz)
+    lenv = zeros(Int, length(sumv))
+
+    for col in cols
+        sumv .= ifelse.(ismissing.(col), sumv, sumv .+ col)
```

Comment:
@nalimilan - do you think we need to do `Statistics._mean_promote` here? I think it should be OK the way we have it now, but these things are always tricky.

Reply:
AFAICT it's not needed when the eltype is concrete, but it is when that's not the case. Note that `zero(Real) === 0` so that can happen even when `zero` is defined. Also, to match `mean`, accumulation has to be done using the return eltype, i.e. above do `zi = zero(eltype(col))/0`.